### PR TITLE
Pull IMigrationTool out of the IMigratableModel interface

### PR DESCRIPTION
**Pull request by ChumpChief**
The customer is expected to implement the IMigratableModel interface with their model in order to use the migration flow.  Currently, part of this requirement is including and providing the migration tool.  However, we can provide this on the customer's behalf with a slightly-decorated container runtime instantiation helper, making it easier to use the migration flow.  This is the core of this PR - removing that requirement on the customer.

This PR does a couple other things too:
1. Switch from container.close() to container.dispose() - this was clarified more recently, and is what we actually want (to fully clean up resources as we no longer need the container data).
2. Switch away from aqueduct-style ContainerRuntimeFactory, instead providing a helper function (a ContainerRuntime.loadRuntime() analog).  This is preferable to avoid class inheritance, and provide more flexibility in how the customer builds their runtime factory.

With these changes, the utils folder is starting to get a little bloated as we have multiple approaches to several of its concepts.  I expect in a subsequent change I'll do some reorganizing of files/directories to clean this up.

[AB#12604](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/12604)

### CODE REVIEW

File: examples/utils/example-utils/src/migrationInterfaces/migratableModel.ts [hunk 1]
```diff
@@ -47,13 +45,11 @@ export interface IMigratableModel
 	extends IVersionedModel,
 		IImportExportModel<unknown, unknown> {
 	/**
-	 * The tool that will be used to facilitate the migration.
-	 */
-	readonly migrationTool: IMigrationTool;
```
Comment (ChumpChief) [inline examples/utils/example-utils/src/migrationInterfaces/migratableModel.ts:52]: This is the main win - customers don't have to provide the IMigrationTool anymore when implementing the IMigratableModel interface.

File: examples/utils/example-utils/src/modelLoader/instantiateMigratableRuntime.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+const migrationToolId = "migration-tool";
+
+const migrationToolRegistryKey = "migration-tool";
+const migrationToolFactory = new MigrationToolFactory();
+
+/**
+ * @internal
+ */
+export const instantiateMigratableRuntime = async <ModelType>(
```
Comment (ChumpChief) [inline examples/utils/example-utils/src/modelLoader/instantiateMigratableRuntime.ts:49]: This helper is used in place of ModelContainerRuntimeFactory by the customer.

File: examples/version-migration/schema-upgrade/src/modelVersion1/containerCode.ts [hunk 3]
```diff
@@ -3,80 +3,82 @@
...
-	}
-
-	/**
-	 * {@inheritDoc ModelContainerRuntimeFactory.containerHasInitialized}
-	 */
-	protected async containerHasInitialized(runtime: IContainerRuntime) {
-		// Force the MigrationTool to instantiate in all cases.  The Quorum it uses must be loaded and running in
-		// order to respond with accept ops, and without this call the MigrationTool won't be instantiated on the
-		// summarizer client.
-		await getDataStoreEntryPoint(runtime, migrationToolId);
```
Comment (ChumpChief) [inline examples/version-migration/schema-upgrade/src/modelVersion1/containerCode.ts:69]: This is an example of why it's nice to do on the customer's behalf - this "always instantiate" requirement would be easy to get wrong.  Now the customer doesn't have to think about it.

Note that now this entire container code can remain ignorant to migration except for using the special helper function to instantiate the migratable runtime.

File: examples/version-migration/schema-upgrade/src/modelVersion1/containerCode.ts [hunk 4]
```diff
@@ -3,80 +3,82 @@
...
-
-const migrationToolRegistryKey = "migration-tool";
-const migrationToolFactory = new MigrationToolFactory();
 
 /**
  * @internal
  */
-export class InventoryListContainerRuntimeFactory extends ModelContainerRuntimeFactory<IInventoryListAppModel> {
+export class InventoryListContainerRuntimeFactory implements IRuntimeFactory {
+	public get IRuntimeFactory() {
```
Comment (ChumpChief) [inline examples/version-migration/schema-upgrade/src/modelVersion1/containerCode.ts:30]: Now that the factory isn't a ModelContainerRuntimeFactory, there's a little more boilerplate, but the customer gains control of lifecycle and inheritance as they see fit.

File: examples/version-migration/schema-upgrade/src/modelVersion1/appModel.ts [hunk 5]
```diff
@@ -25,7 +25,6 @@ export class InventoryListAppModel implements IInventoryListAppModel, IMigratabl
 
 	public constructor(
 		public readonly inventoryList: IInventoryList,
-		public readonly migrationTool: IMigrationTool,
```
Comment (ChumpChief) [inline examples/version-migration/schema-upgrade/src/modelVersion1/appModel.ts:28]: The app model also gets to become ignorant of the migration tool - now it's only responsible for providing import/export, version, and dispose.

File: examples/version-migration/schema-upgrade/src/modelVersion1/appModel.ts [hunk 6]
```diff
@@ -61,7 +60,7 @@ export class InventoryListAppModel implements IInventoryListAppModel, IMigratabl
 		return `version:one\n${inventoryItemStrings.join("\n")}`;
 	};
 
-	public close() {
-		this.container.close();
+	public dispose() {
```
Comment (tyler-cai-microsoft) [inline examples/version-migration/schema-upgrade/src/modelVersion1/appModel.ts:63]: This seems like a breaking change of sorts

Comment (ChumpChief) [reply examples/version-migration/schema-upgrade/src/modelVersion1/appModel.ts:63]: I suppose folks could implement IMigratableModel.dispose() differently here if they want (i.e. have it call container.close() instead if they're supporting old loader).

Should just follow the exact same pattern as whatever they do today for deciding when to call container.close() vs. container.dispose() in their non-migration scenarios.

File: examples/utils/example-utils/src/modelLoader/instantiateMigratableRuntime.ts (hunk 2 above)
Comment (scottn12) [reply examples/utils/example-utils/src/modelLoader/instantiateMigratableRuntime.ts:49]: Thoughts on creating an interface for the inputs here (i.e. `IMigratableRuntimeProps`)?

**Review by scottn12 — APPROVED** [review]
LGTM 🤠

File: examples/utils/example-utils/src/modelLoader/instantiateMigratableRuntime.ts (hunk 2 above)
Comment (ChumpChief) [reply examples/utils/example-utils/src/modelLoader/instantiateMigratableRuntime.ts:60]: Could be good longer term, though for initial iteration it's a little easier to just use vanilla parameters.  I'll add a TODO here to consider as we finalize.

File: examples/utils/example-utils/src/modelLoader/instantiateMigratableRuntime.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+import { NamedFluidDataStoreRegistryEntries } from "@fluidframework/runtime-definitions/internal";
+
+import {
+	getDataStoreEntryPoint,
+	type IMigrationTool,
+	MigrationToolFactory,
+} from "../index.js";
+
+/**
+ * @internal
```
[the remark this answers is not in the document]
Comment (ChumpChief) [reply examples/utils/example-utils/src/modelLoader/instantiateMigratableRuntime.ts:27]: Added in latest

File: examples/utils/example-utils/src/migrator/migrator.ts [hunk 8]
```diff
@@ -10,32 +10,42 @@ import { assert } from "@fluidframework/core-utils/internal";
...
 	IMigratableModel,
+	IMigrationTool,
 	IMigrator,
 	IMigratorEvents,
 	MigrationState,
 } from "../migrationInterfaces/index.js";
-import type { IDetachedModel, IModelLoader } from "../modelLoader/index.js";
+import type { IDetachedModel, IMigratableModelLoader } from "../modelLoader/index.js";
+
+interface MigratableParts {
```
[the remark this answers is not in the document]
Comment (ChumpChief) [reply examples/utils/example-utils/src/migrator/migrator.ts:25]: Added in latest

**Review by ChumpChief — COMMENTED** [review]
Added requested docs plus a few more in places I touched - there's still more documentation needed but I'll tackle it more exhaustively as the project progresses.

**Pull request merged**